bin/propolis-server/src/lib/server.rs

```diff
@@ -318,36 +320,111 @@ fn instance_spec_from_request(
     Ok(VersionedInstanceSpec::V0(spec_builder.finish()))
 }
 
-/// Attempts to register an Oximeter server reporting metrics from a new
-/// instance, returning the producer registry from that server on success.
-async fn register_oximeter(
-    server_context: &DropshotEndpointContext,
-    cfg: &MetricsEndpointConfig,
-    vm_id: uuid::Uuid,
+/// Register an Oximeter server reporting metrics from a new instance.
+///
+/// This spawns a tokio task which will indefinitely attempt to register with
+/// Nexus as an Oximeter metric producer. Once the registration succeeds, it
+/// will also arrange for the statistics produced for the managed instance to be
+/// collected.
+async fn register_oximeter_in_background(
+    services: Arc<ServiceProviders>,
+    cfg: MetricsEndpointConfig,
+    registry: ProducerRegistry,
+    virtual_machine: VirtualMachine,
     log: Logger,
-) -> anyhow::Result<ProducerRegistry> {
-    assert!(server_context.services.oximeter_stats.lock().await.is_none());
-    assert!(server_context
-        .services
-        .oximeter_server_task
-        .lock()
+) {
+    assert!(services.oximeter_stats.lock().await.is_none());
+    assert!(services.oximeter_server_task.lock().await.is_none());
```

Comment:
Maybe use `replace` when doing the assignments, and assert then that the old values were none?  (Or both, for the ol' belt-n-suspenders)

Reply:
Kept these asserts, and added them on the value returned from `Option::replace` as well, in d61fd69.